<a href="https://colab.research.google.com/github/lalit1511/Maths-Stats-using-python/blob/main/Dot_Products_withoutcode.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![](https://betterexplained.com/wp-content/uploads/2012/02/dot_product_components.png)

![](https://www.google.com/url?sa=i&url=https%3A%2F%2Fwww.cuemath.com%2Falgebra%2Fdot-product%2F&psig=AOvVaw1VmXz4YwiM8u56f0WmXdeO&ust=1646985300054000&source=images&cd=vfe&ved=0CAsQjRxqFwoTCLi2hqGJu_YCFQAAAAAdAAAAABAb)

![](https://d138zd1ktt9iqe.cloudfront.net/media/seo_landing_files/matrix-representation-of-dot-product-1626103121.png)

# **`WATCH VIDEOS IN THE PORTAL`**
## **`Watch Video 1: Dot Product(Theory)`**

### **CHeck this blog for more on dot product theory+code**
https://towardsdatascience.com/dot-product-in-linear-algebra-for-data-science-using-python-f113fb60c8f#:~:text=This%20matrix%20multiplication%20is%20also%20called%20Dot%20Product.&text=Important%20Rule%3A%20You%20can%20only,rows%20of%20the%20second%20matrix.


# **TASK**

We will be using a fabricated dataset of user ratings for the movies to learn

- how we use vector operations in the data science(extracting, finding magnitude, finding angle between vectors)
-  how to do dot product in coding and learn to find similarity between user ratings to a certain movie
- Learn to find list of movies based on user preference
- We will also sort list of movies based on their popularity
- We will also recommend movies based on a given similar movie

# Task 1 - Load and study the data

Notice these things: Load the data and study its features such as:
- The number of users
- The number of movies
- The ranges of ratings

In [ ]:
# Load "numpy" and "pandas" for manipulating numbers, vectors and data frames
import numpy as np
import pandas as pd


In [ ]:
# Read in the "User_Movie_Ratings.csv" file as a Pandas Data Frame
# Note: Make sure the code and the data are in the same folder or specify the appropriate path for the data
df = pd.read_csv('/content/User_Movie_Ratings.csv')
df



User  Thor  Justice League  Spiderman  The Notebook  Casablanca  \
0      Arnold     6               5          6             4           5   
1         Ben     9              10          8             3           1   
2   Catherine    10               8          9             5           6   
3       Disha     6               7          7            10           9   
4      Euclid     7               4          3             5           8   
5     Farhaan     6               5          9             4           4   
6      Gaurav     8               8          9             3           1   
7     Hilbert     6               5          8             6           5   
8      Indira    10               7          9             5           6   
9       Jason    10               9          9             7           4   
10      Kenny     3               2          5             9           8   
11    Lambert     5               6          4            10           9   
12     Manish     7               6          6             5           6   
13       Niti     3               3          2             7           8   
14     Oliver     4               5          4             2           2   
15     Pranav     8               9          8             4           1   
16      Qader     8               7          8             5           4   
17       Ravi     8               7          9            10           9   
18    Sushant     5               8          6             1           1   
19    Tavneet     1               1          4             7           6   

    Beauty and the Beast  Dolittle  The Kissing Booth  Annabelle  Insidious  
0                      4         9                  8          2          1  
1                      1         2                  5          7          8  
2                      7         8                  5          9          9  
3                     10         7                  6          5          6  
4                      5         9                 10          8          7  
5                      7         9                  9          5          4  
6                      1         7                  5          7          7  
7                      8        10                  9          2          1  
8                      4         2                  1          6          8  
9                      6         1                  1          5          4  
10                     8         4                  1          1          3  
11                     8         1                  2         10          9  
12                     8         7                  6          6          5  
13                     7        10                  9          4          5  
14                     1         7                  6          1          2  
15                     2         3                  1          8          7  
16                     6         3                  4          6          5  
17                     8         8                  8          9         10  
18                     2         7                  8          6          5  
19                     8         9                 10          1          5

Feature Description:
- The data set contains ratings given by users to movies on a scale of 1 to 10.
- Each row contains the ratings given to all the movies by a particular user.
- Each column contains the ratings given to a particular movie by all the users.

In [ ]:
# Take a brief look at the data frame using ".head()"
df.head()


User  Thor  Justice League  Spiderman  The Notebook  Casablanca  \
0     Arnold     6               5          6             4           5   
1        Ben     9              10          8             3           1   
2  Catherine    10               8          9             5           6   
3      Disha     6               7          7            10           9   
4     Euclid     7               4          3             5           8   

   Beauty and the Beast  Dolittle  The Kissing Booth  Annabelle  Insidious  
0                     4         9                  8          2          1  
1                     1         2                  5          7          8  
2                     7         8                  5          9          9  
3                    10         7                  6          5          6  
4                     5         9                 10          8          7

In [ ]:
# Check the dimensions of the data frame using ".shape"

df.shape

(20, 11)

In [ ]:
# View basic information about the data frame using ".info()"

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   User                  20 non-null     object
 1   Thor                  20 non-null     int64 
 2   Justice League        20 non-null     int64 
 3   Spiderman             20 non-null     int64 
 4   The Notebook          20 non-null     int64 
 5   Casablanca            20 non-null     int64 
 6   Beauty and the Beast  20 non-null     int64 
 7   Dolittle              20 non-null     int64 
 8   The Kissing Booth     20 non-null     int64 
 9   Annabelle             20 non-null     int64 
 10  Insidious             20 non-null     int64 
dtypes: int64(10), object(1)
memory usage: 1.8+ KB


In [ ]:
# View a statistical summary of the data frame using ".describe()"
# Note: Use ".transpose()" to make the summary easier to read
df.describe().transpose()


count  mean       std  min   25%  50%   75%   max
Thor                   20.0  6.50  2.502630  1.0  5.00  6.5  8.00  10.0
Justice League         20.0  6.10  2.381950  1.0  5.00  6.5  8.00  10.0
Spiderman              20.0  6.65  2.300458  2.0  4.75  7.5  9.00   9.0
The Notebook           20.0  5.60  2.643761  1.0  4.00  5.0  7.00  10.0
Casablanca             20.0  5.15  2.852054  1.0  3.50  5.5  8.00   9.0
Beauty and the Beast   20.0  5.55  2.855742  1.0  3.50  6.5  8.00  10.0
Dolittle               20.0  6.15  3.116594  1.0  3.00  7.0  9.00  10.0
The Kissing Booth      20.0  5.70  3.180533  1.0  3.50  6.0  8.25  10.0
Annabelle              20.0  5.40  2.817240  1.0  3.50  6.0  7.25  10.0
Insidious              20.0  5.55  2.605157  1.0  4.00  5.0  7.25  10.0

# Observations

- There are 20 rows and 10 columns in the data
- Each row corresponds to the movie ratings given by a particular user
- Each column corresponds to the user ratings given to a particular movie
- The movie ratings range from 1 to 10 and they are whole numbers

# Task 2 - Extract vectors from the data

![](https://www.grc.nasa.gov/www/k-12/airplane/Images/vectors.jpg)

### **Scalars, Vectors, Matrices in Data Science**

![](https://miro.medium.com/max/957/1*jRyyMAhS_NZxqyv3EoLCvg.png)

By Looking at the df.head() in previous code, you have seen that there are two types of vectors that we can extract from the data:

- Row vectors - these contain the movie ratings given by a particular user
- Column vectors - these contain the user ratings given to a particular movie

## **`Watch Video 2 : Dot Product(Coding)`**

### **This is a nice blog on the use of iloc and loc methods**

https://towardsdatascience.com/how-to-use-loc-and-iloc-for-selecting-data-in-pandas-bd09cb4c3d79

**NOTE**: open in incognito mode if you are not able to read.

In [ ]:
# Access the movie ratings vector for the user "Ben" using ".loc
df.loc['Ben']

KeyError: ignored

In [ ]:
# Access the movie ratings vector for the user "Niti" using ".loc[]"
df.loc['niti']

KeyError: ignored

In [ ]:
# Access the user ratings vector for the movie "Thor"
# Note: You may access column vectors either directly or using ".loc[]"

df['Thor']

0      6
1      9
2     10
3      6
4      7
5      6
6      8
7      6
8     10
9     10
10     3
11     5
12     7
13     3
14     4
15     8
16     8
17     8
18     5
19     1
Name: Thor, dtype: int64

In [ ]:
# Access the user ratings vector for the movie "Beauty and the Beast"
df.['Beauty and the Beast']



SyntaxError: ignored

# Observations

- There are two types of vectors that we can work with:
- The movie ratings vectors for a particular user (row vectors) that are 10-long since there are 10 movies in the data set
- The user ratings vectors for a particular movie (column vectors) that are 20-long since there are 20 users in the data set

# Task 3 - Calculate dot products of vectors

Check this documentation for dot product

https://numpy.org/doc/stable/reference/generated/numpy.dot.html

In [ ]:
# Calculate the dot product of the movie ratings vectors for the users "Disha" and "Oliver" using "np.dot()"
# Note: You must enter two vectors as parameters to "np.dot()"
np.dot(df.loc['Disha'],df.loc['Oliver'])


KeyError: ignored

In [ ]:
# Calculate the dot product of the movie ratings vector for the user "Disha" with itself using "np.dot()"
np.dot(df.loc['Disha'],df.loc['Disha'])


KeyError: ignored

In [ ]:
# Calculate the dot product of the user ratings vectors for the movies "Casablanca" and "Annabelle" using "np.dot()"
np.dot(df['Casablanca'],df['Annabelle'])


566

In [ ]:
# Calculate the dot product of the user ratings vector for the movie "Casablanca" with itself using "np.dot()"
np.dot(df['Casablanca'],df['Casablanca'])


685

# Observations

- Dot products can only be calculated for vectors of the same length
- Dot products between movie ratings vectors for a particular user in this data can theoretically range from 10 to 1000
- Dot products between user ratings vectors for a particular movie in this data can theoretically range from 20 to 2000

# Task 4 - Create a function to calculate magnitudes of vectors

### We will create a function to calculate the magnitude of a vector and use it as a measure of the strength of that vector

![](https://www.onlinemathlearning.com/image-files/magnitude-vector.png)

In [ ]:
# Create a function called "mag()" which takes in a vector and returns its magnitude
# Note: The square root of the dot product of a vector with itself is equal to its magnitude
# Note: Use the "np.dot()" function to calculate dot products and the "np.sqrt()" function to calculate square roots
def mag(vec):
  return(np.sqrt(np.dot(vec,vec)))



In [ ]:
# Use the function "mag()" to find the magnitude of the movie ratings vector for the user "Pranav"
mag(df.loc['Pranav'])



KeyError: ignored

In [ ]:
# Use the function "mag()" to find the magnitude of the user ratings vector for the movie "Justice League"
mag(df['Justice League'])



29.189039038652847

# Observations

- The magnitude of a vector indicates the strength of the values in that vector
- The smaller the magnitude of a vector, the weaker the vector, and vice versa
- The larger the magnitude of a vector, the stronger the vector, and vice versa
- The magnitudes of the movie ratings vectors for a particular user in this data can theoretically range from 3.16 to 31.6
- The magnitudes of the user ratings vectors for a particular movie in this data can theoretically range from 4.47 to 44.7

# Task 5 - Print a sorted list of users in terms of how much they like movies in general

### **NOTE**
- The magnitude of a vector indicates the strength of the values in that vector
- The smaller the magnitude of a vector, the weaker the vector and vice versa
- The larger the magnitude of a vector, the stronger the vector and vice versa
- We can use the magnitudes of the movie ratings vectors given by particular users to measure their interest in movies

In [ ]:
# Calculate the magnitudes of the movie ratings vectors for all 20 users and sort them in descending order
# Note: Use the "mag()" function to calculate magnitudes of movie ratings vectors for particular users
# Note: Use these magnitudes as a measure of interest of users in watching movies
# Note: Store the resulting values in a Pandas Series with index as user names and name the series "users"
# Note: The "index" parameter of the series can be set as "df.index"
# Note: You may need to specify the "dtype" parameter of the series as "float64" to avoid some warnings
# Note: Use the ".sort_values()" function with the "ascending" parameter set to "False"
users =pd.Series(index=df.index,dtype='float64')

for user in users.index:
  users.loc[user]=mag(df.loc[user])

    # save mag of user from df in users
    


users=users.sort_values(ascending=False) # sort here

TypeError: ignored

In [ ]:
# Print the "users" series




User
Ravi         27.349589
Catherine    24.617067
Disha        23.685439
Lambert      22.538855
Euclid       21.954498
Hilbert      20.880613
Farhaan      20.639767
Indira       20.297783
Niti         20.149442
Jason        20.149442
Ben          19.949937
Gaurav       19.798990
Manish       19.798990
Tavneet      19.339080
Pranav       18.788294
Qader        18.439089
Sushant      17.464249
Arnold       17.435596
Kenny        16.552945
Oliver       12.489996
dtype: float64

# Observations

- The user "Ravi" has the highest interest in watching movies
- The user "Oliver" has the least interest in watching movies

# Task 6 - Print a sorted list of movies in terms of their popularity

### **NOTE**

- The magnitude of a vector indicates the strength of the values in that vector
- The smaller the magnitude of a vector, the weaker the vector, and vice versa
- The larger the magnitude of a vector, the stronger the vector, and vice versa
- We can use the magnitudes of the user ratings vectors given to particular movies to measure their popularity

In [ ]:
# Calculate the magnitudes of the user ratings vectors for all 10 movies and sort them in descending order
# Note: Use the "mag()" function to calculate magnitudes of user ratings vectors for particular movies
# Note: Use these magnitudes as a measure of the popularity of the movies
# Note: Store the resulting values in a Pandas Series with index as movie names and name the series "movies"
# Note: The "index" parameter of the series can be set as "df.columns"
# Note: You may need to specify the "dtype" parameter of the series as "float64" to avoid some warnings
# Note: Use the ".sort_values()" function with the "ascending" parameter set to "False"

movies = # code here
for movie in movies.index:
    # save mag for each movie in movies
    

# sort by values    
movies = 

In [ ]:
# Print the "movies" series


Spiderman               31.384710
Thor                    31.048349
Dolittle                30.675723
Justice League          29.189039
The Kissing Booth       29.017236
Beauty and the Beast    27.766887
The Notebook            27.568098
Insidious               27.294688
Annabelle               27.092434
Casablanca              26.172505
dtype: float64

# Observations

- The most popular movie is "Spiderman"
- The least popular movie is "Casablanca"

# Task 7 - Create a function to calculate the angle between two vectors

### **Some Mathematical Insights**:

- The angle between any two vectors can be used as a similarity measure between those two vectors.  We will calculate the angle between two vectors using dot products

- The algorithm to find the angle between two vectors is as follows:
1. Calculate the dot product of the two vectors
2. Calculate the magnitudes of the two vectors
3. Divide the value obtained in step 1 by the product of the values obtained in step 2
4. Find the inverse cosine of the value obtained in step 3

### **Note:**
- Note: The angle obtained will be in radians
- Note: You may code this out step by step or in one line

In [ ]:
# Create a function called "ang()" which takes in two vectors and returns the angle between them
# Note: Use "np.dot()" to calculate dot products
# Note: Use "mag()" to calculate magnitudes
# Note: Use "np.arccos()" to calculate inverse cosines
def ang(vec1, vec2):
    # code here
    num=np.dot(vec1,vec2)
    mag1=mag(vec1)
    mag2=mag(vec2)
    den=mag1*mag2
    angle=np.arccos(num/den)
    return(angle)

    

In [ ]:
# Calculate the angle between the movie ratings vectors for the users "Ben" and "Ravi" using the function "ang()"
ang(df['Ben'],df['Ravi'])


KeyError: ignored

In [ ]:
# Calculate the angle between the user ratings vectors for the movies "Spiderman" and "The Notebook" using the function "ang()"
ang(df["Spiderman" ],df["The Notebook"])



0.566621652026982

In [ ]:
# Calculate the angle between the user ratings vector for the movieang(df["Spiderman" ],df["The Notebook"]) and itself using the function "ang()"
ang(df[ ],df["The Notebook"])


<ipython-input-45-cdf1e810d758>:10: RuntimeWarning: invalid value encountered in arccos
  angle = np.arccos(num/den)


nan

- Notice that you get a "nan" value, that is, "not a number"
- Let's calculate the angle between the user ratings vector for the movie "Annabelle" and itself step by step

In [ ]:
# First, store the user ratings vector for the movie "Annabelle" in two variables "u" and "v"
u = 
v = 

In [ ]:
# Next, calculate the numerator of the angle formula, that is the dot product of "u" and "v" and store the value as "num"
num = 

In [ ]:
# Print the value of the numerator
num

734

In [ ]:
# Next, calculate the magnitudes of the two vectors "u" and "v" and store them in two variables "mag_u" and "mag_v"
mag_u = 
mag_v = 

In [ ]:
# Next, calculate the denominator of the angle formula, that is the product of "mag_u" and "mag_v" and store the value as "den"
den = 


In [ ]:
# Print the value of the denominator



733.9999999999999

- Notice that the denominator is nearly equal to the numerator but slightly less than it. This is because of the precision limitation when storing numbers using Python.

In [ ]:
# Calculate the inverse cosine of the ratio of the numerator to the denominator



<ipython-input-57-1e6125f2c95e>:2: RuntimeWarning: invalid value encountered in arccos
  np.arccos(num/den)


nan

- Since this ratio (734 / 733.9999999999999) is greater than 1, no such inverse cosine angle is defined. This can be handled in this case by rounding the denominator.
- The values in the vectors (and hence their magnitudes) are large enough that rounding them to whole numbers will work
- In case you are working with values where rounding is not an option, you will need to use conditional statements

In [ ]:
# Redefine the function "ang()" by rounding the denominator part of the angle formula
def ang(vec1, vec2):
    # code here

In [ ]:
# Calculate the angle between the user ratings vector for the movie "Annabelle" and itself using the function "ang()"



0.0

# Observations

- The dot product of two vectors can be used to calculate the angle between the two vectors
- The angle obtained is in radians
- The angle between two vectors measures the similarity between those vectors
- The angle between a vector and itself is 0 radians
- The smaller the angle between two vectors, the higher their similarity score, and vice versa
- The larger the angle between two vectors, the lower their similarity score, and vice versa
- To overcome Python's precision limitations, we round the denominator of the angle formula

# Task 8 - Create a function that returns the name of the person who is most similar to a given user

- The angle between two vectors can be used as a measure of similarity between the two vectors
- The smaller the angle between two vectors, the higher their similarity score, and vice versa
- The larger the angle between two vectors, the lower their similarity score, and vice versa
- We will create a function that returns the name of the person who is most similar to a given user

In [ ]:
# Define a function "usersim()" that takes in the name of a user and returns the name of the user most similar to them
# Note: Use the "ang()" function to measure similarities between users (using their movie ratings vectors)
# Note: You may create a temporary Pandas Series within the "usersim()" function to store the angle values
# Note: The "index" parameter of the series can be set as "df.index"
# Note: You may need to specify the "dtype" parameter of the series as "float64" to avoid some warnings
# Note: You may sort the entries in this series and return the second index entry
# Note: Use the ".sort_values()" function with the default value for the "ascending" parameter, which is "True"
# Note: The first entry after sorting will be trivial
def usersim(user1):
    # code here

In [ ]:
# Use the function "usersim()" to find the name of the user who is most similar to the user "Manish"



'Ravi'

In [ ]:
# Use the function "usersim()" to find the name of the user who is most similar to the user "Tavneet"




'Niti'

In [ ]:
# Use the function "usersim()" to find the name of the user who is most similar to the user "Indira"



'Qader'

In [ ]:
# Use the function "usersim()" to find the name of the user who is most similar to the user "Qader"



'Catherine'

# Observations

- The angle between the movie ratings vectors for two particular users is a measure of similarity between the users
- The similarity score is not symmetric
- Rather it pertains to a specific case (such as a specific user)
- This is because it is possible for X to be most similar to Y and Y to be most similar to Z
- This happens when X and Z are not as similar to each other as X and Y or Y and Z
- Transitivity, on the other hand, is a useful assumption for recommendations

# Task 9 - Create a function that returns the name of the movie that is most similar to a given movie

### **Logic**

- The angle between two vectors can be used as a measure of similarity between the two vectors. The smaller the angle between two vectors, the higher their similarity score, and vice versa.
- The larger the angle between two vectors, the lower their similarity score, and vice versa

**What's Next ?**- We will create a function that returns the name of the movie which is most similar to a given movie

In [ ]:
# Define a function "movesim()" that takes in the name of a movie and returns the name of the movie most similar to it
# Note: Use the "ang()" function to measure similarities between movies (using their user ratings vectors)
# Note: You may create a temporary Pandas Series within the "moviesim()" function to store the angle values
# Note: The "index" parameter of the series can be set as "df.columns"
# Note: You may need to specify the "dtype" parameter of the series as "float64" to avoid some warnings
# Note: You may sort the entries in this series and return the second index entry
# Note: Use the ".sort_values()" function with the default value for the "ascending" parameter, which is "True"
# Note: The first entry after sorting will be trivial
def moviesim(movie1):
    # code here

In [ ]:
# Use the function "moviesim()" to find the name of the movie that is most similar to the movie "Casablanca"



'The Notebook'

In [ ]:
# Use the function "moviesim()" to find the name of the movie that is most similar to the movie "Insidious"



'Annabelle'

In [ ]:
# Use the function "moviesim()" to find the name of the movie that is most similar to the movie "Spiderman"



'Thor'

In [ ]:
# Use the function "moviesim()" to find the name of the movie that is most similar to the movie "Thor"



'Justice League'

# Observations

- The angle between the user ratings vectors for two particular movies is a measure of similarity between the movies
- The similarity score is not symmetric
- Rather it pertains to a specific case (such as a specific movie)
- This is because it is possible for X to be most similar to Y and Y to be most similar to Z
- This happens when X and Z are not as similar to each other as X and Y or Y and Z
- Transitivity, on the other hand, is a useful assumption for recommendations

# Conclusions

- From the user movie ratings data, we can extract movie ratings vectors and user ratings vectors. We can use magnitudes of movie ratings vectors for particular users to measure their interest in movies.
- We can use magnitudes of user ratings vectors for particular movies to measure their popularity.
- We can use the angle between movies ratings vectors of two users to measure their similarity.
- We can use the angle between user ratings vectors of two movies to measure their similarity.
- Using these measures, we can recommend similar movies to similar users

 **This is a very basic look into the working methodology of recommender systems.**

# **FEEDBACK**

https://forms.zohopublic.in/cloudyml/form/CloudyMLStatisticsFeedbackForm/formperma/WV946wnf0sDM_tOlH87RxZR9yMceKWGrtuPOyXzzCRc